src/clients/storage/StorageClientBase-inl.h

```diff
@@ -120,180 +74,135 @@ StorageClientBase<ClientType, ClientManagerType>::collectResponse(
-  using TransportException = apache::thrift::transport::TTransportException;
-  auto context = std::make_shared<ResponseContext<Request, RemoteFunc, Response>>(
-      requests.size(), std::move(remoteFunc));
-
-  DCHECK(!!ioThreadPool_);
-
-  for (auto& req : requests) {
-    auto& host = req.first;
-    auto spaceId = req.second.get_space_id();
-    auto res = context->insertRequest(host, std::move(req.second));
-    DCHECK(res.second);
-    evb = ioThreadPool_->getEventBase();
-    // Invoke the remote method
-    folly::via(evb, [this, evb, context, host, spaceId, res]() mutable {
-      auto client = clientsMan_->client(host, evb, false, FLAGS_storage_client_timeout_ms);
-      // Result is a pair of <Request&, bool>
-      auto start = time::WallClock::fastNowInMicroSec();
-      context
-          ->serverMethod(client.get(), *res.first)
-          // Future process code will be executed on the IO thread
-          // Since all requests are sent using the same eventbase, all
-          // then-callback will be executed on the same IO thread
-          .via(evb)
-          .thenValue([this, context, host, spaceId, start](Response&& resp) {
-            auto& result = resp.get_result();
-            bool hasFailure{false};
-            for (auto& code : result.get_failed_parts()) {
-              VLOG(3) << "Failure! Failed part " << code.get_part_id() << ", failed code "
-                      << static_cast<int32_t>(code.get_code());
-              hasFailure = true;
-              context->resp.emplaceFailedPart(code.get_part_id(), code.get_code());
-              if (code.get_code() == nebula::cpp2::ErrorCode::E_LEADER_CHANGED) {
-                auto* leader = code.get_leader();
-                if (isValidHostPtr(leader)) {
-                  updateLeader(spaceId, code.get_part_id(), *leader);
-                } else {
-                  invalidLeader(spaceId, code.get_part_id());
+  std::vector<folly::Future<StatusOr<Response>>> respFutures;
+  respFutures.reserve(requests.size());
+
+  auto hosts = std::make_shared<std::vector<HostAddr>>(requests.size());
+  auto totalLatencies = std::make_shared<std::vector<int32_t>>(requests.size());
+
+  for (const auto& req : requests) {
+    auto start = time::WallClock::fastNowInMicroSec();
+
+    size_t i = respFutures.size();
+    (*hosts)[i] = req.first;
+    // Future process code will be executed on the IO thread
+    // Since all requests are sent using the same eventbase, all
+    // then-callback will be executed on the same IO thread
+    auto fut = getResponse(evb, req.first, req.second, std::move(remoteFunc))
+                   .ensure([totalLatencies, i, start]() {
+                     (*totalLatencies)[i] = time::WallClock::fastNowInMicroSec() - start;
+                   });
+
+    respFutures.emplace_back(std::move(fut));
+  }
+
+  return folly::collectAll(respFutures)
+      .deferValue([this, requests = std::move(requests), totalLatencies, hosts](
```

Comment:
Do we need to add `via` after `collectAll` and before `deferValue`? I can't recall that which thread would trigger the callback in `defer`.

My point is the rpc of `respFutures` will run on the parameter `evb`, what if `collectAll` is running on same `evb`, can we make sure that `rpc1, rpc2, ..., collectAll, rpcn` won't happen? This would cause deadlock

Reply:
If we call the via for the SemiFuture of collectAll, these tasks will be executed immediately on evb. that's not our expected result. We want to let user of this interface to specify which thread these tasks work on. It's same as the original logics.

Reply:
👌